### PR TITLE
add option for mean and std to be tuples

### PR DESCRIPTION
**Pull request by JoanFM**
### Description
I propose to allow the `user` to instantiate `Normalize` object with `mean` and `std` as tuples.

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### CODE REVIEW

File: kornia/augmentation/augmentation.py [hunk 1]
```diff
@@ -35,8 +32,6 @@
     adjust_contrast,
     adjust_saturation,
     adjust_hue,
-    adjust_gamma,
-    Invert,
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:39]: you shouldn't remove those

File: kornia/augmentation/augmentation.py [hunk 2]
```diff
@@ -1090,8 +1085,8 @@ class Normalize(AugmentationBase2D):
     Where `mean` is :math:`(M_1, ..., M_n)` and `std` :math:`(S_1, ..., S_n)` for `n` channels,
 
     Args:
-        mean (torch.Tensor): Mean for each channel.
-        std (torch.Tensor): Standard deviations for each channel.
+        mean (Union[torch.Tensor, Tuple[float], float]): Mean for each channel.
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:1088]:
```suggestion
        mean (Union[torch.Tensor, Tuple[float, ...], float]): Mean for each channel.
```

File: kornia/augmentation/augmentation.py [hunk 3]
```diff
@@ -1090,8 +1085,8 @@ class Normalize(AugmentationBase2D):
...
+        std (Union[torch.Tensor, Tuple[float], float]): Standard deviations for each channel.
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:1089]:
```suggestion
        std (Union[torch.Tensor, Tuple[float, ...], float]): Standard deviations for each channel.
```

File: kornia/augmentation/augmentation.py [hunk 4]
```diff
@@ -1106,7 +1101,7 @@ class Normalize(AugmentationBase2D):
     """
 
     def __init__(
-        self, mean: torch.Tensor, std: torch.Tensor,
+        self, mean: Union[torch.Tensor, Tuple[float], float], std: Union[torch.Tensor, Tuple[float], float],
```
Comment (edgarriba) [inline kornia/augmentation/augmentation.py:1104]:
```suggestion
        self, mean: Union[torch.Tensor, Tuple[float, ...], float], std: Union[torch.Tensor, Tuple[float, ...], float],
```

File: kornia/enhance/normalize.py [hunk 5]
```diff
@@ -24,8 +24,8 @@ class Normalize(nn.Module):
     Where `mean` is :math:`(M_1, ..., M_n)` and `std` :math:`(S_1, ..., S_n)` for `n` channels,
 
     Args:
-        mean (Union[torch.Tensor, float]): Mean for each channel.
-        std (Union[torch.Tensor, float]): Standard deviations for each channel.
+        mean (Union[torch.Tensor, Tuple[float], float]): Mean for each channel.
```
Comment (edgarriba) [inline kornia/enhance/normalize.py:27]:
```suggestion
        mean (Union[torch.Tensor, Tuple[float, ...], float]): Mean for each channel.
```

File: kornia/enhance/normalize.py [hunk 6]
```diff
@@ -24,8 +24,8 @@ class Normalize(nn.Module):
...
+        std (Union[torch.Tensor, Tuple[float], float]): Standard deviations for each channel.
```
Comment (edgarriba) [inline kornia/enhance/normalize.py:28]:
```suggestion
        std (Union[torch.Tensor, Tuple[float, ...], float]): Standard deviations for each channel.
```

File: kornia/enhance/normalize.py [hunk 7]
```diff
@@ -45,7 +45,7 @@ class Normalize(nn.Module):
         torch.Size([1, 4, 3, 3])
     """
 
-    def __init__(self, mean: Union[torch.Tensor, float], std: Union[torch.Tensor, float]) -> None:
+    def __init__(self, mean: Union[torch.Tensor, Tuple[float], float], std: Union[torch.Tensor, Tuple[float], float]) -> None:
```
Comment (edgarriba) [inline kornia/enhance/normalize.py:48]:
```suggestion
    def __init__(self, mean: Union[torch.Tensor, Tuple[float, ...], float], std: Union[torch.Tensor, Tuple[float, ...], float]) -> None:
```

File: kornia/enhance/normalize.py [hunk 8]
```diff
@@ -72,8 +72,8 @@ def normalize(
 
     Args:
         data (torch.Tensor): Image tensor of size :math:`(*, C, ...)`.
-        mean (Union[torch.Tensor, float]): Mean for each channel.
-        std (Union[torch.Tensor, float]): Standard deviations for each channel.
+        mean (Union[torch.Tensor, Tuple[float], float]): Mean for each channel.
```
Comment (edgarriba) [inline kornia/enhance/normalize.py:75]:
```suggestion
        mean (Union[torch.Tensor, Tuple[float, ...], float]): Mean for each channel.
```

File: kornia/enhance/normalize.py [hunk 9]
```diff
@@ -72,8 +72,8 @@ def normalize(
...
+        std (Union[torch.Tensor, Tuple[float], float]): Standard deviations for each channel.
```
Comment (edgarriba) [inline kornia/enhance/normalize.py:76]:
```suggestion
        std (Union[torch.Tensor, Tuple[float, ...], float]): Standard deviations for each channel.
```

File: test/augmentation/test_augmentation.py [hunk 10]
```diff
@@ -2461,11 +2460,18 @@ def test_smoke(self, device, dtype):
...
         assert str(f) == repr
 
-    def test_random_normalize(self, device, dtype):
-        f = Normalize(mean=torch.tensor([1.]), std=torch.tensor([.5]), p=1., return_transform=True)
-        f1 = Normalize(mean=torch.tensor([1.]), std=torch.tensor([.5]), p=0., return_transform=True)
-        f2 = Normalize(mean=torch.tensor([1.]), std=torch.tensor([.5]), p=1.)
-        f3 = Normalize(mean=torch.tensor([1.]), std=torch.tensor([.5]), p=0.)
+    @pytest.mark.parametrize("mean, std",
+                             [
+                                 1., .5,
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:2465]:
```suggestion
                                 (1., .5,),
```

File: test/augmentation/test_augmentation.py [hunk 11]
```diff
@@ -2461,11 +2460,18 @@ def test_smoke(self, device, dtype):
...
+                                 (1.,), (.5, ),
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:2466]:
```suggestion
                                 ((1.,), (.5, )),
```

File: test/augmentation/test_augmentation.py [hunk 12]
```diff
@@ -2461,11 +2460,18 @@ def test_smoke(self, device, dtype):
...
+                                 torch.tensor([1.]), torch.tensor([.5])
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:2467]:
```suggestion
                                 (torch.tensor([1.]), torch.tensor([.5]))
```

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: kornia/augmentation/augmentation.py (hunk 1 above)
Comment (JoanFM) [reply kornia/augmentation/augmentation.py:39]: ok, just rracting to my IDE saying were not used

File: test/augmentation/test_augmentation.py (hunk 11 above)
Comment (edgarriba) [inline test/augmentation/test_augmentation.py]:
```suggestion
                                 ((1.,), (.5, )),
                                 ((1., 2., 3., 4.,), (4., 3., 2., 1.,))
```

**Review by edgarriba — COMMENTED** [review]
add this extra test

**Review by shijianjian — CHANGES_REQUESTED** [review]
Please remove those prints first. 

In principle, I think it is better to keep all functional functions in a tensor only setting, with less tensor initializations. The tuple might be provided as the syntax sugar for nn.Modules, in which tensors might be initialised and stored properly in the module.

**Comment by JoanFM** [comment]
> Please remove those prints first.
> 
> In principle, I think it is better to keep all functional functions in a tensor only setting, with less tensor initializations. The tuple might be provided as the syntax sugar for nn.Modules, in which tensors might be initialised and stored properly in the module.

Sure, prints are leftover, because I am not sure how, but my test was not going to the function I expected. And I still not understand.

I agree `functions` can work with `torch.Tensors` but `initialization` should support regular python tuples, I think it feels more natural to provide `mean` or `std` as regular python objects than with `torch.Tensor`, why to put such burden in the user end?

But yes, adding it as `syntax` sugar in the `nn` module discussed in the issue #520 would work for me

File: kornia/enhance/normalize.py [hunk 13]
```diff
@@ -99,6 +99,14 @@ def normalize(
     if isinstance(std, float):
         std = torch.tensor([std] * shape[1], device=data.device, dtype=data.dtype)
 
+    if isinstance(mean, tuple):
```
Comment (JoanFM) [inline kornia/enhance/normalize.py:102]: Hey @shijianjian .

What I think it would make sense is to review everywhere in the code and allow only to provide `mean` and `std` as `torch.Tensor` at this stage.

So the conversion from whatever accepted input to `torch.Tensor` is in every `instance constructor`.

Like this you can skip all these checks and conversions at every functional iteration, and keep the code more mantainable

Comment (JoanFM) [reply kornia/enhance/normalize.py:102]: Actually, I am not sure, are these functions that the user should be able to use from their own code (are they public) or are they to only be used by the internal `kornia` objects (therefore private)?

If so, I think this strategy should be followed in general and code will be better unittested maybe?

**Comment by shijianjian** [comment]
> I agree `functions` can work with `torch.Tensors` but `initialization` should support regular python tuples, I think it feels more natural to provide `mean` or `std` as regular python objects than with `torch.Tensor`, why to put such burden in the user end?
> 

I don't think it is a good design to convert a tuple or list or a number to torch tensor in a functional call. It firstly reduced the efficiency whilst init and destroy small tensors. I think I have discussed with @edgarriba at some point about this, that to keep the Kornia core functions to accept purely tensors.

> 
> But yes, adding it as `syntax` sugar in the `nn` module discussed in the issue #520 would work for me

The syntax sugar I meant is to support different data types like tuple and lists in modulized implementations. In which way the mean and std will only be generated at the initialisation stage and will be reused afterwards. 

In short, functional functions support tensor only, while the nn.Module implementation is the more user friendly choice.

**Comment by JoanFM** [comment]
> > I agree `functions` can work with `torch.Tensors` but `initialization` should support regular python tuples, I think it feels more natural to provide `mean` or `std` as regular python objects than with `torch.Tensor`, why to put such burden in the user end?
> 
> I don't think it is a good design to convert a tuple or list or a number to torch tensor in a functional call. It firstly reduced the efficiency whilst init and destroy small tensors. I think I have discussed with @edgarriba at some point about this, that to keep the Kornia core functions to accept purely tensors.
> 
> > But yes, adding it as `syntax` sugar in the `nn` module discussed in the issue #520 would work for me
> 
> The syntax sugar I meant is to support different data types like tuple and lists in modulized implementations. In which way the mean and std will only be generated at the initialisation stage and will be reused afterwards.
> 
> In short, functional functions support tensor only, while the nn.Module implementation is the more user friendly choice.

Sorry, I missunderstood.

I totallly agree, would it be good if I do the `conversion` at `init` level? even for float `std` and `mean`? That would involve adding an optional parameter `num_channels` in case your std or mean are `floats`.

Would that make sense? I think is an effort that makes sense to avoid anything not `tensors` in the functional calls.

File: kornia/enhance/normalize.py (hunk 13 above)
Comment (shijianjian) [reply kornia/enhance/normalize.py:102]: > Hey @shijianjian .
> 
> 
> 
> What I think it would make sense is to review everywhere in the code and allow only to provide `mean` and `std` as `torch.Tensor` at this stage.
> 
> 
> 
> So the conversion from whatever accepted input to `torch.Tensor` is in every `instance constructor`.
> 
> 
> 
> Like this you can skip all these checks and conversions at every functional iteration, and keep the code more mantainable 

Correct. That is exactly what I meant. You only think about tensors for functional calls, and provide some easy API with nn.Module.

**Comment by shijianjian** [comment]
> 
> I totallly agree, would it be good if I do the `conversion` at `init` level? even for float `std` and `mean`? That would involve adding an optional parameter `num_channels` in case your std or mean are `floats`.
> 
Yes. All tensors shall be prepared in __init__. I would discard float option. 

> 
> Would that make sense? I think is an effort that makes sense to avoid anything not `tensors` in the functional calls.

Yes. Much better.

**Comment by JoanFM** [comment]
> > I totallly agree, would it be good if I do the `conversion` at `init` level? even for float `std` and `mean`? That would involve adding an optional parameter `num_channels` in case your std or mean are `floats`.
> 
> Yes. All tensors shall be prepared in **init**. I would discard float option.
> 
> > Would that make sense? I think is an effort that makes sense to avoid anything not `tensors` in the functional calls.
> 
> Yes. Much better.

Will try to do when I have time this for `normalize`. Maybe it can serve as a good starting point for maybe other parts that may suffer from the same problem?

**Comment by shijianjian** [comment]
> Will try to do when I have time this for `normalize`. Maybe it can serve as a good starting point for maybe other parts that may suffer from the same problem?

Yes. It is one legacy issue in kornia.

**Comment by shijianjian** [comment]
@JoanFM hi, how's things going?

**Comment by JoanFM** [comment]
> @JoanFM hi, how's things going?

not much time lately

**Comment by JoanFM** [comment]
Hey @shijianjian ,

I tried to do the changes we discussed, can u check if they make sense? Then I can maybe add better testing. I would however would like to know if there is some formatting script that I can run to avoid breaking the format as I did in this PR (used my IDE formatting)

**Comment by shijianjian** [comment]
Yes. That is exactly it. I used VSCode and each tab is 4 spaces in my settings. Probably your IDE is using 8 spaces?

One thing I am thinking is to avoid:
```python
    mean = torch.as_tensor(mean, device=data.device, dtype=data.dtype)
    std = torch.as_tensor(std, device=data.device, dtype=data.dtype)
```
If the user input tensors in different devices and types, we shall let it crash. We shall only take care of our internal created tensors. @JoanFM @edgarriba what do you think?

**Comment by edgarriba** [comment]
@shijianjian I have another concern about this function in here: https://github.com/kornia/kornia/pull/987/files#diff-b88d696bfd3e23924d367439e3084df1fe473a9c9d1f0688bd9252fa220c6997R126

I think that in case of a tensor is provided, should also be the use that makes sure about the shape can be broadcasted. At most, we could support that we apply the mean and std to the last two image dimensions (..., H, W)

@lferraz found some issues too while using normalise

**Comment by shijianjian** [comment]
Seems to be an easy fix. Just to update ````data.shape[-3]``` to ````data.shape[1]```.

```python
    if mean.shape and mean.shape[0] != 1:
        if mean.shape[0] != data.shape[1] and mean.shape[:2] != data.shape[:2]:
            raise ValueError(f"mean length and number of channels do not match. Got {mean.shape} and {data.shape}.")
    # Allow broadcast on channel dimension
    if std.shape and std.shape[0] != 1:
        if std.shape[0] != data.shape[1] and std.shape[:2] != data.shape[:2]:
            raise ValueError(f"std length and number of channels do not match. Got {std.shape} and {data.shape}.")
    mean = torch.as_tensor(mean, device=data.device, dtype=data.dtype)
    std = torch.as_tensor(std, device=data.device, dtype=data.dtype)
    if mean.shape:
        mean = mean[..., :, None]
    if std.shape:
        std = std[..., :, None]
    out: torch.Tensor = (data.view(shape[0], shape[1], -1) - mean) / std
    return out.view(shape)
```

**Comment by edgarriba** [comment]
@JoanFM could you try @shijianjian suggestion ?

**Comment by JoanFM** [comment]
> @JoanFM could you try @shijianjian suggestion ?

@edgarriba I will as soon as I get some time. Any suggestion as per how I can easily fix the formatting issues?

**Comment by edgarriba** [comment]
@JoanFM I'll help on that during the weekend

**Comment by JoanFM** [comment]
@edgarriba I have applied the proposed changes. Any suggestion on how to adapt the formatting problem? and what about the testing?

**Review by shijianjian — CHANGES_REQUESTED** [review]
Could you please:
1. Checkout the ```augmentation.py``` and ```test_augmenation.py``` to the previous state without modification.
2. Use a text editor (e.g. vscode, sublime, or even vim) to write the tests.
3. Please test both functional and modulized operations.

**Comment by JoanFM** [comment]
> Could you please:
> 
> 1. Checkout the `augmentation.py` and `test_augmenation.py` to the previous state without modification.
> 2. Use a text editor (e.g. vscode, sublime, or even vim) to write the tests.
> 3. Please test both functional and modulized operations.

Hey @shijianjian ,

I did change the formatting, but now I am not sure where each of the tests should live in

File: kornia/enhance/normalize.py [hunk 14]
```diff
@@ -45,10 +45,22 @@ class Normalize(nn.Module):
...
 
         super(Normalize, self).__init__()
 
+        if isinstance(mean, float):
+            mean = torch.tensor([mean] * num_channels)
+
+        if isinstance(std, float):
+            std = torch.tensor([std] * num_channels)
+
+        if isinstance(mean, tuple):
```
Comment (shijianjian) [inline kornia/enhance/normalize.py:58]: We should allow list as well.

File: kornia/augmentation/augmentation.py [hunk 15]
```diff
@@ -1106,11 +1106,25 @@ class Normalize(AugmentationBase2D):
     """
 
     def __init__(
-        self, mean: torch.Tensor, std: torch.Tensor,
+        self, mean: Union[torch.Tensor, Tuple[float, ...], float],
+        std: Union[torch.Tensor, Tuple[float, ...], float],
+        num_channels: Optional[int] = 3,
```
Comment (shijianjian) [inline kornia/augmentation/augmentation.py:1111]: What if tensor gives 4 while num_channels got only 3. A bit confusing. I would remove this arg.

Probably use torch.expand(mean, input.shape[1]) when len(mean.shape) == 0.

**Review by shijianjian — CHANGES_REQUESTED** [review]
Put those tests there is okay to me. Besides, can you update the denormolize as well?

File: kornia/enhance/normalize.py [hunk 16]
```diff
@@ -45,10 +45,22 @@ class Normalize(nn.Module):
         torch.Size([1, 4, 3, 3])
     """
 
-    def __init__(self, mean: Union[torch.Tensor, float], std: Union[torch.Tensor, float]) -> None:
+    def __init__(self, mean: Union[torch.Tensor, Tuple[float, ...], float], std: Union[torch.Tensor, Tuple[float, ...], float], num_channels: Optional[int]) -> None:
 
         super(Normalize, self).__init__()
 
+        if isinstance(mean, float):
+            mean = torch.tensor([mean] * num_channels)
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply kornia/enhance/normalize.py:53]: yes, but that's why is not set as an optional parameter.

File: kornia/augmentation/augmentation.py (hunk 15 above)
Comment (JoanFM) [reply kornia/augmentation/augmentation.py:1111]: the point is to avoid any extra computations when function is called.

Comment (shijianjian) [reply kornia/augmentation/augmentation.py:1112]: I wouldn't encourage this if it is conflicting with other params. 

Meanwhile, torch.expand will not consume much and do not require the user to know their input in the first place.

I am okay with both solutions, either solve the confusion or with torch.expand.

Comment (JoanFM) [reply kornia/augmentation/augmentation.py:1112]: ok will use torch expand then

**Review by edgarriba — CHANGES_REQUESTED** [review]
@JoanFM check the mypy - something went wrong

**Comment by edgarriba** [comment]
@JoanFM you just need to rebase from master to fix the latest pytorch 1.9 formatting issues

File: test/augmentation/test_augmentation.py [hunk 17]
```diff
@@ -2603,6 +2603,27 @@ def test_smoke(self, device, dtype):
...
+        "mean, std", [((1.0, 1.0, 1.0), (0.5, 0.5, 0.5)), (1.0, 0.5), (torch.tensor([1.0]), torch.tensor([0.5]))]
+    )
+    def test_random_normalize_different_parameter_types(mean, std):
+        f = Normalize(mean=mean, std=std, p=1)
+        data = torch.ones(2, 3, 256, 313)
+        if isinstance(mean, float):
+            expected = (data - torch.tensor(mean)) / torch.tensor(std)
+        else:
+            expected = (data - torch.tensor(mean[0])) / torch.tensor(std[0])
+        assert_allclose(f(data), expected)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:2617]:
```suggestion
        assert_close(f(data), expected)
```

File: test/enhance/test_normalize.py [hunk 18]
```diff
@@ -117,6 +117,27 @@ def test_module(self, device, dtype):
...
+        "mean, std", [((1.0, 1.0, 1.0), (0.5, 0.5, 0.5)), (1.0, 0.5), (torch.tensor([1.0]), torch.tensor([0.5]))]
+    )
+    def test_random_normalize_different_parameter_types(mean, std):
+        f = kornia.enhance.Normalize(mean=mean, std=std)
+        data = torch.ones(2, 3, 256, 313)
+        if isinstance(mean, float):
+            expected = (data - torch.tensor(mean)) / torch.tensor(std)
+        else:
+            expected = (data - torch.tensor(mean[0])) / torch.tensor(std[0])
+        assert_allclose(f(data), expected)
```
Comment (edgarriba) [inline test/enhance/test_normalize.py:131]:
```suggestion
        assert_close(f(data), expected)
```

**Review by edgarriba — APPROVED** [review]
